pommerman/network/ion_client/network.py

```diff
@@ -0,0 +1,186 @@
+    and Server"""
+
+    def __init__(self, ip):
+        """Arguments:  
+        * ip: The IP of the server"""
+        self.ws_ = websocket.create_connection(
+            "ws://" + str(ip), enable_multithread=True
+        )
+        self.lock = threading.Lock()
+
+    def check(self):
+        """Description: Check if the server actually responds"""
+        self.lock.acquire()
+        try:
+            self.ws_.send(
+                rapidjson.dumps({
+                    "intent": constants.NetworkCommands.check.value
+                })
+            )
+            message_recieved = rapidjson.loads(self.ws_.recv())
+            self.lock.release()
+            if message_recieved["intent"
+                               ] == constants.NetworkCommands.status_ok.value:
+                return [
+                    message_recieved["players"], message_recieved["matches"]
+                ]
+            elif message_recieved[
+                "intent"
+            ] == constants.NetworkCommands.status_full.value:
+                raise Exception(constants.Exceptions.net_server_full.value)
+            raise Exception(constants.Exceptions.net_respond_fail.value)
+        except:
+            self.lock.release()
+            raise Exception(constants.Exceptions.net_connect_fail.value)
+
+    def join_list(self, room=False):
+        """Argument:  
+        * room: The room to be created/joined. If False, the public room will \
+be joined
+        Description: Check if the server actually responds"""
+        self.lock.acquire()
+        if not room:
+            self.ws_.send(
+                rapidjson.dumps(
+                    {
+                        "intent": constants.NetworkCommands.match.value
+                    }
+                )
+            )
+        else:
+            self.ws_.send(
+                rapidjson.dumps(
+                    {
+                        "intent": constants.NetworkCommands.room.value,
+                        "room": room
+                    }
+                )
+            )
+        self.lock.release()
+        decoded_msg = rapidjson.loads(self.ws_.recv())
+        if (
+            decoded_msg["intent"] == constants.NetworkCommands.status_full.value
+        ):
+            if room:
+                raise Exception(constants.Exceptions.room_full.value)
+            else:
+                raise Exception(constants.Exceptions.match_full.value)
+        self.id = decoded_msg["player_id"]
+        self.mode = decoded_msg["mode"]
+
+    def waitm(self):
+        """Description: Wait for a response from the server regarding a match 
+        request"""
+        self.lock.acquire()
+        try:
+            r = rapidjson.loads(self.ws_.recv())
+            if r["intent"] == constants.NetworkCommands.match_start.value:
+                self.match_id = r["match_id"]
+        except:
+            self.lock.release()
+            raise Exception(constants.Exceptions.net_server_closed.value)
```

Comment:
Are you absolutely sure that that is what has happened here? If it's not a net_server_closed, then it's going to be really confusing. An alternative could be to just catch the exception and then say what that exception is (with the try/catch being so that we can release the lock).

Reply:
I replaced them with `finally` and it's a much much better solution.